src/compiler/ir/README.md

```diff
@@ -0,0 +1,162 @@
+# General structure
+
+The IR has a fairly standard structure, made up of basic blocks. Each basic
+block has a list of statement and a terminator. While the implementation
+occasionally refers to "phi nodes", the IR uses basic block arguments rather
+than phi nodes.
+
+The IR only represents the body of methods. The "shape" of the program, eg. its
+classes, fields, method names and signatures, ... are all stored in the AST
+instead. For each `Method` node in the AST, a corresponding `FnAnalysis` is
+created, which, among others, contains the `MethodIR`.
+
+A `MethodIR` is composed of one or more `FunctionIR`. This distinction is used
+to encode the bodies of `when` clauses that appear within the method. The first
+`FunctionIR` corresponds to the method body itself, whereas subsequent blocks
+correspond to when block bodies.
+
+Once created by the `ir/builder` pass, the IR is immutable. All new information
+is represented in external side structures, as part of the `FnAnalysis`
+structure.
+
+# Type information
+
+The IR itself contains almost no type information. Only types which appear in
+the method body at the source level is included in the IR. At the moment, this
+is just types used in a match block.
+
+The inferred type of IR values is stored in maps in `FnAnalysis`. The type of
+`v` in basic block `bb` can be found in `analysis->typecheck->types[bb][v]`.
+
+Similarily, at construction, every method call site is assigned a
+`TypeArgumentsId`, used to identify the call's type arguments. Given an `id`,
+the inferred type arguments can be found in
+`analysis->typecheck->type_arguments[id]`.
+
+In both cases however, these are the polymorphic types. If a method or class is
+generic, these types may mention type variables. The non-generic types will
+only be known after reachabilitiy and reification.
+
+# Statements
+## `NewStmt`
+
+Allocates a new object of a given class. It may optionally be given a `parent`
+which designates the region in which the object should be allocated. If no
+parent is specified, the object is allocated in a new region.
```

Comment:
It depends whether we consider this an optimisation or an inference. In the former case, programs would be correct without that pass, and I'm happy for this to happen in MLIR (it would be a "merge elision" pass). If it's an inference pass, programs can't type check without that pass, so we.

The truth is we'll need both. Region inference is necessary to get certain programs to compile, but "merge elision" is more powerful (it can take advantage of inlining and other generic optimizations, for example).